crates/agglayer-prover/src/executor.rs

```diff
@@ -0,0 +1,255 @@
+};
+use tokio::task::spawn_blocking;
+use tower::{
+    limit::ConcurrencyLimitLayer, timeout::TimeoutLayer, util::BoxCloneService, Service,
+    ServiceBuilder, ServiceExt,
+};
+use tracing::error;
+
+#[cfg(test)]
+mod tests;
+
+/// ELF of the pessimistic proof program
+const ELF: &[u8] =
+    include_bytes!("../../pessimistic-proof-program/elf/riscv32im-succinct-zkvm-elf");
+
+#[derive(Clone)]
+pub struct Executor {
+    network: BoxCloneService<Request, Response, Error>,
+    local: BoxCloneService<Request, Response, Error>,
+}
+
+impl Executor {
+    pub fn build_network_service<S>(
+        timeout: Duration,
+        service: S,
+    ) -> BoxCloneService<Request, Response, Error>
+    where
+        S: Service<Request, Response = Response, Error = Error> + Send + Clone + 'static,
+        <S as Service<Request>>::Future: std::marker::Send,
+    {
+        BoxCloneService::new(
+            ServiceBuilder::new()
+                .layer(TimeoutLayer::new(timeout))
+                .service(service)
+                .map_err(|error| match error.downcast::<Error>() {
+                    Ok(error) => *error,
+                    Err(error) => Error::ProverFailed(error.to_string()),
+                }),
+        )
+    }
+
+    pub fn build_local_service<S>(
+        timeout: Duration,
+        concurrency: usize,
+        service: S,
+    ) -> BoxCloneService<Request, Response, Error>
+    where
+        S: Service<Request, Response = Response, Error = Error> + Send + Clone + 'static,
+        <S as Service<Request>>::Future: std::marker::Send,
+    {
+        BoxCloneService::new(
+            ServiceBuilder::new()
+                .layer(TimeoutLayer::new(timeout))
+                .layer(ConcurrencyLimitLayer::new(concurrency))
+                .service(service)
+                .map_err(|error| match error.downcast::<Error>() {
+                    Ok(error) => *error,
+                    Err(error) => Error::ProverFailed(error.to_string()),
+                }),
+        )
+    }
+
+    #[cfg(test)]
+    pub fn new_with_services(
+        network: BoxCloneService<Request, Response, Error>,
+        local: BoxCloneService<Request, Response, Error>,
+    ) -> Self {
+        Self { network, local }
+    }
+
+    pub fn new(config: &ProverConfig) -> Self {
+        let network_prover = NetworkProver::new();
+        let (_proving_key, verification_key) = network_prover.setup(ELF);
+        let network = Self::build_network_service(
+            config.network_prover.proving_timeout,
+            NetworkExecutor {
+                prover: Arc::new(network_prover),
+                verification_key,
+            },
+        );
+
+        let prover = CpuProver::new();
+        let (proving_key, verification_key) = prover.setup(ELF);
+
+        let local = Self::build_local_service(
+            config.cpu_prover.proving_timeout,
+            config.cpu_prover.max_concurrency_limit,
```

Comment:
I'm wondering, will the timeout also free up resources? Ideally, a timeout on local prover would somehow kill the proving process to free up the CPU. Similarly, for network prover, it should cancel the request provided the service provides such facility.

Reply:
For the `CpuProver`, when the timeout occurs the future returns and stop, but as SP1 `CpuProver` doesn't support async, we can't really stop the task when launched. I didn't find any workaround for that :/ 

For the `NetworkProver`, when the timeout occurs, we cut the connection. I don't know how SP1 is reacting to that.

Reply:
> For the `CpuProver`, when the timeout occurs the future returns and stop, but as SP1 `CpuProver` doesn't support async, we can't really stop the task when launched. I didn't find any workaround for that :/

Yea not trivial. The best I can think of is to run the prover workload in a separate process and just kill it on timeout. Could make an useful command line tool as well.